### ISSUE TITLE
CMake: Windows test and shared builds

### ISSUE THREAD

**Issue by vtraag** [igraph/igraph#1528]
On Windows there are a couple of tests that we decided we should disable. Some of them don't seem to use any internal functions, yet don't compile correctly on Windows. 

Let us take one example `igraph_random_sample`. Building that tests errors out with the error: `error LNK2019: unresolved external symbol __imp_igraph_random_sample referenced in function error_test `. This suggest that the function is not properly exported/imported (i.e. incorrectly marked with `dllspec`). However, it is perfectly well defined here:
https://github.com/igraph/igraph/blob/a4f2fec3ea25348cf7fed1507ae47381b1673ea9/include/igraph_nongraph.h#L76-L77
I even double checked whether `DECLDIR` is correctly defined when including this header, and it is.

Now here is where it gets weird. If I instead include these two lines in `igraph_random.h` instead of in `igraph_nongraph.h`, it does compile correctly! You might think it is then a problem with the whole file `igraph_nongraph.h`, but that is also incorrect. For example, `igraph_nongraph.h` also contains `igraph_fisher_yates_shuffle`, but that compiles perfectly fine with the tests.

**Comment by vtraag** [comment]
I think I actually also just found out the problem: `igraph_nongraph.h` is not properly included in `random.c`.

**Comment by vtraag** [comment]
There are two options of how we can solve this: either we move the relevant declarations to `igraph_random.h`, or alternatively, we include also `igraph_nongraph.h` in `random.c`. At any rate, we should clean that up.

Possibly similar problem appear in other parts of the codebase. Would there be any proper way of detecting these?

**Comment by vtraag** [comment]
It actually turns out `igraph_fisher_yates_shuffle` is nowehere implemented. The test file `igraph_fisher_yates_shuffle.c` does not use that implementation. We should hence just remove the declaration of `igraph_fisher_yates_shuffle`.

**Comment by szhorvat** [comment]
I am fixing many of these issues right now, will push in 15 min.

**Comment by szhorvat** [comment]
I fixed a bunch of these. Can you test again @vtraag ?

Here's the method I used:

 - disable `-Werror` and enable [`-Wmissing-prototypes`](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wmissing-prototypes)
 - record the warnings, and filter out those which refer to functions with `_i_` in the name
 - manually sort through the rest and add any missing includes

The nastiest bug was a mismatch between the prototype and definition of `igraph_hrg_sample`.

**Comment by vtraag** [comment]
I think it would be better to separate `igraph_nongraph.h`. It is now a very strange include file that just collects all sort of random stuff. Wouldn't it be better to split it into the relevant parts? That is, we have `other.c` which includes most of these functions, so we can easily add `zeroin` to that file. We then only need to add `igraph_random_sample` to `igraph_random.h`, where, frankly, it would also better fit I think? Finally, we are then left with `bfgs`, for which we would then need a separate `igraph_bfgs,h` header file, although we could also think of including it in `other.c`.

By the way @tamas: the function `igraph_power_law_fit` is also included in `other.c`, but I though that power-law fitting was done in `plfit/plfit.c`?

We can then leave the rest as is, because that will be part of a bigger restructuring exercise anyway.

**Comment by szhorvat** [comment]
Sure, but I would suggest to first fix compilation issues. Then when everything is working, we can refactor some things.

**Comment by ntamas** [comment]
+1 for getting rid of `igraph_nongraph.h` after the compilation / linking issues are fixed.

As for `igraph_power_law_fit` -- it is in `other.c` because the `src/plfit` folder is more-or-less a vendored copy of the relevant parts of my [plfit repo](https://github.com/ntamas/plfit) so it is not supposed to contain any igraph-specific stuff.

**Comment by vtraag** [comment]
> first fix all compilation and linking issues. Then when everything is working, we can refactor (or better organize) some things.

Fair enough. 

I'm now testing, and have now included the following tests:
```
igraph_get_eids
igraph_eccentricity
igraph_get_all_simple_paths
igraph_random_sample
```

These tests just compiled correctly.

Any other tests that I should include?

> As for `igraph_power_law_fit` -- it is in `other.c` because the `src/plfit` folder is more-or-less a vendored copy of the relevant parts of my [plfit repo](https://github.com/ntamas/plfit) so it is not supposed to contain any igraph-specific stuff.

Ah right, that explains that.

**Comment by szhorvat** [comment]
When I checked, it seemed to me that _most_ disabled tests shouldn't be disabled.

I would suggest enabling every test whose name does not suggest that it concerns an internal function. Then if any of them fail, check the test file and see if the reason is that an internal function is used.  If it is, I would add a comment in the cmake file,  noting which internal function it uses.

After a quick glance, I would NOT enable these (and enable all the rest):

```
    2wheap
    biguint
    d_indheap
    igraph_hashtable
    igraph_i_cutheap
    igraph_marked_queue
    igraph_set
    igraph_sparsemat2
    igraph_trie

    igraph_i_layout_sphere

    igraph_bfs     <-- uses igraph_i_bfs
```

**Comment by szhorvat** [comment]
It really looks like we should make a 0.8.4 when all this stuff is fixed ... I was hoping that 0.8.3 would be the last one, but the cmake work revealed so many issues that I think it's fair to issue a 0.8.4 release after the cmake branch has been merged (we might find more problems so it's worth waiting until then).

**Comment by vtraag** [comment]
OK, I simply renabled everything (i.e. set `BUILD_INTERNAL_TESTS` to `YES) to check everything. The ones that still fail are:

```
test_igraph_scg_grouping (2wheap)
test_igraph_fisher_yates_shuffle (igraph_rewire, igraph_qsort_r)
test_igraph_power_law_fit (igraph_set)
test_igraph_maximum_bipartite_matching (igraph_marked_queue)
test_igraph_layout_merge2 (igraph_sparsemat2)
test_igraph_layout_lgl (igraph_layout_merge)
test_igraph_sparsemat7 (igraph_layout_davidson_harel)
test_igraph_grg_game (igraph_i_cutheap)
test_igraph_strvector (igraph_i_layout_sphere)
test_igraph_feedback_arc_set (igraph_hashtable)
test_igraph_correlated_game (igraph_trie)
test_igraph_community_fastgreedy (igraph_all_st_cuts)
test_igraph_betweenness (igraph_bfs)
test_pajek_bipartite (d_indheap)
test_biguint_betweenness (biguint)
```

We have discussed previously if we should make some of these functions properly "public", as they also include internal functions (`igraph_i*`).

> It really looks like we should make a 0.8.4 when all this stuff is fixed

I am not sure if this would be necessary, but we can discuss about that. It might be quite a lot of work to cherry-pick some corrections without making API breaking changes.

**Comment by szhorvat** [comment]
>  It might be quite a lot of work to cherry-pick some corrections

Absolutely don't cherry pick! I meant releasing the fixes that are already on the master branch.

**Comment by szhorvat** [comment]
Can you please quote the exact error message once more for a few of these failures?  The error should mention where the functions were referenced. I simply don't see these functions used in the respective test programs.

> test_igraph_fisher_yates_shuffle (igraph_rewire, igraph_qsort_r)

Neither of these two functions are used _directly_ in `igraph_fisher_yates_shuffle.c`

**Comment by szhorvat** [comment]
Actually I don't understand the `a (b)` notation at all. `a`s are names of test executables, but `b`s seem to be names of unrelated _tests_ (i.e. _not_ function names)?

**Comment by vtraag** [comment]
Sorry, my bad, I misinterpreted the log files. These are the remaining test failures, including exact function references:

```
2wheap.obj igraph_2wheap_init referenced in function main [tests\test_2wheap.vcxproj]
2wheap.obj igraph_2wheap_destroy referenced in function main [tests\test_2wheap.vcxproj]
2wheap.obj igraph_2wheap_push_with_index referenced in function main [tests\test_2wheap.vcxproj]
2wheap.obj igraph_2wheap_empty referenced in function main [tests\test_2wheap.vcxproj]
2wheap.obj igraph_2wheap_size referenced in function main [tests\test_2wheap.vcxproj]
2wheap.obj igraph_2wheap_max referenced in function main [tests\test_2wheap.vcxproj]
2wheap.obj igraph_2wheap_get referenced in function main [tests\test_2wheap.vcxproj]
2wheap.obj igraph_2wheap_delete_max referenced in function main [tests\test_2wheap.vcxproj]
2wheap.obj igraph_2wheap_delete_max_index referenced in function main [tests\test_2wheap.vcxproj]
2wheap.obj igraph_2wheap_modify referenced in function main [tests\test_2wheap.vcxproj]
2wheap.obj igraph_2wheap_check referenced in function main [tests\test_2wheap.vcxproj]
igraph_rewire.obj igraph_rewire_core referenced in function check_rewiring [tests\test_igraph_rewire.vcxproj]
igraph_qsort_r.obj __imp_igraph_qsort_r referenced in function main [tests\test_igraph_qsort_r.vcxproj]
igraph_set.obj igraph_set_init referenced in function main [tests\test_igraph_set.vcxproj]
igraph_set.obj igraph_set_destroy referenced in function main [tests\test_igraph_set.vcxproj]
igraph_set.obj igraph_set_empty referenced in function main [tests\test_igraph_set.vcxproj]
igraph_set.obj igraph_set_clear referenced in function main [tests\test_igraph_set.vcxproj]
igraph_set.obj igraph_set_size referenced in function main [tests\test_igraph_set.vcxproj]
igraph_set.obj igraph_set_add referenced in function main [tests\test_igraph_set.vcxproj]
igraph_set.obj igraph_set_contains referenced in function main [tests\test_igraph_set.vcxproj]
igraph_set.obj igraph_set_iterate referenced in function print_set [tests\test_igraph_set.vcxproj]
igraph_marked_queue.obj igraph_marked_queue_init referenced in function main [tests\test_igraph_marked_queue.vcxproj]
igraph_marked_queue.obj igraph_marked_queue_destroy referenced in function main [tests\test_igraph_marked_queue.vcxproj]
igraph_marked_queue.obj igraph_marked_queue_empty referenced in function main [tests\test_igraph_marked_queue.vcxproj]
igraph_marked_queue.obj igraph_marked_queue_iselement referenced in function main [tests\test_igraph_marked_queue.vcxproj]
igraph_marked_queue.obj igraph_marked_queue_push referenced in function main [tests\test_igraph_marked_queue.vcxproj]
igraph_marked_queue.obj igraph_marked_queue_start_batch referenced in function main [tests\test_igraph_marked_queue.vcxproj]
igraph_marked_queue.obj igraph_marked_queue_pop_back_batch referenced in function main [tests\test_igraph_marked_queue.vcxproj]
igraph_sparsemat2.obj igraphdgemv_ referenced in function igraph_matrix_dgemv [tests\test_igraph_sparsemat2.vcxproj]
igraph_layout_merge.obj igraph_i_layout_mergegrid_init referenced in function main [tests\test_igraph_layout_merge.vcxproj]
igraph_layout_merge.obj igraph_i_layout_mergegrid_destroy referenced in function main [tests\test_igraph_layout_merge.vcxproj]
igraph_layout_merge.obj igraph_i_layout_merge_place_sphere referenced in function main [tests\test_igraph_layout_merge.vcxproj]
igraph_layout_merge.obj igraph_i_layout_merge_dla referenced in function main [tests\test_igraph_layout_merge.vcxproj]
igraph_layout_davidson_harel.obj igraph_i_segments_intersect referenced in function intersect [tests\test_igraph_layout_davidson_harel.vcxproj]
igraph_layout_davidson_harel.obj igraph_i_point_segment_dist2 referenced in function distance [tests\test_igraph_layout_davidson_harel.vcxproj]
igraph_i_cutheap.obj igraph_i_cutheap_init referenced in function main [tests\test_igraph_i_cutheap.vcxproj]
igraph_i_cutheap.obj igraph_i_cutheap_destroy referenced in function main [tests\test_igraph_i_cutheap.vcxproj]
igraph_i_cutheap.obj igraph_i_cutheap_empty referenced in function main [tests\test_igraph_i_cutheap.vcxproj]
igraph_i_cutheap.obj igraph_i_cutheap_popmax referenced in function main [tests\test_igraph_i_cutheap.vcxproj]
igraph_i_cutheap.obj igraph_i_cutheap_update referenced in function main [tests\test_igraph_i_cutheap.vcxproj]
igraph_i_layout_sphere.obj igraph_i_layout_sphere_2d referenced in function main [tests\test_igraph_i_layout_sphere.vcxproj]
igraph_i_layout_sphere.obj igraph_i_layout_sphere_3d referenced in function main [tests\test_igraph_i_layout_sphere.vcxproj]
igraph_hashtable.obj igraph_hashtable_init referenced in function main [tests\test_igraph_hashtable.vcxproj]
igraph_hashtable.obj igraph_hashtable_destroy referenced in function main [tests\test_igraph_hashtable.vcxproj]
igraph_hashtable.obj igraph_hashtable_addset referenced in function main [tests\test_igraph_hashtable.vcxproj]
igraph_hashtable.obj igraph_hashtable_addset2 referenced in function main [tests\test_igraph_hashtable.vcxproj]
igraph_hashtable.obj igraph_hashtable_get referenced in function main [tests\test_igraph_hashtable.vcxproj]
igraph_hashtable.obj igraph_hashtable_getkeys referenced in function main [tests\test_igraph_hashtable.vcxproj]
igraph_hashtable.obj igraph_hashtable_reset referenced in function main [tests\test_igraph_hashtable.vcxproj]
igraph_trie.obj igraph_trie_init referenced in function main [tests\test_igraph_trie.vcxproj]
igraph_trie.obj igraph_trie_destroy referenced in function main [tests\test_igraph_trie.vcxproj]
igraph_trie.obj igraph_trie_get referenced in function main [tests\test_igraph_trie.vcxproj]
igraph_trie.obj igraph_trie_check referenced in function main [tests\test_igraph_trie.vcxproj]
igraph_trie.obj igraph_trie_idx referenced in function main [tests\test_igraph_trie.vcxproj]
igraph_trie.obj igraph_trie_size referenced in function main [tests\test_igraph_trie.vcxproj]
igraph_all_st_cuts.obj igraph_marked_queue_init referenced in function main [tests\test_igraph_all_st_cuts.vcxproj]
igraph_all_st_cuts.obj igraph_marked_queue_destroy referenced in function main [tests\test_igraph_all_st_cuts.vcxproj]
igraph_all_st_cuts.obj igraph_marked_queue_push referenced in function main [tests\test_igraph_all_st_cuts.vcxproj]
igraph_all_st_cuts.obj igraph_marked_queue_start_batch referenced in function main [tests\test_igraph_all_st_cuts.vcxproj]
igraph_all_st_cuts.obj igraph_estack_init referenced in function main [tests\test_igraph_all_st_cuts.vcxproj]
igraph_all_st_cuts.obj igraph_estack_destroy referenced in function main [tests\test_igraph_all_st_cuts.vcxproj]
igraph_all_st_cuts.obj igraph_estack_push referenced in function main [tests\test_igraph_all_st_cuts.vcxproj]
igraph_all_st_cuts.obj igraph_i_all_st_cuts_pivot referenced in function main [tests\test_igraph_all_st_cuts.vcxproj]
igraph_bfs.obj igraph_i_bfs referenced in function main [tests\test_igraph_bfs.vcxproj]
d_indheap.obj igraph_d_indheap_init referenced in function main [tests\test_d_indheap.vcxproj]
d_indheap.obj igraph_d_indheap_destroy referenced in function main [tests\test_d_indheap.vcxproj]
d_indheap.obj igraph_d_indheap_empty referenced in function main [tests\test_d_indheap.vcxproj]
d_indheap.obj igraph_d_indheap_push referenced in function main [tests\test_d_indheap.vcxproj]
d_indheap.obj igraph_d_indheap_max referenced in function main [tests\test_d_indheap.vcxproj]
d_indheap.obj igraph_d_indheap_delete_max referenced in function main [tests\test_d_indheap.vcxproj]
d_indheap.obj igraph_d_indheap_size referenced in function main [tests\test_d_indheap.vcxproj]
d_indheap.obj igraph_d_indheap_reserve referenced in function main [tests\test_d_indheap.vcxproj]
d_indheap.obj igraph_d_indheap_max_index referenced in function main [tests\test_d_indheap.vcxproj]
biguint.obj igraph_biguint_init referenced in function main [tests\test_biguint.vcxproj]
biguint.obj igraph_biguint_destroy referenced in function main [tests\test_biguint.vcxproj]
biguint.obj igraph_biguint_set_limb referenced in function main [tests\test_biguint.vcxproj]
biguint.obj igraph_biguint_equal referenced in function main [tests\test_biguint.vcxproj]
biguint.obj igraph_biguint_inc referenced in function main [tests\test_biguint.vcxproj]
biguint.obj igraph_biguint_dec referenced in function main [tests\test_biguint.vcxproj]
biguint.obj igraph_biguint_add referenced in function main [tests\test_biguint.vcxproj]
biguint.obj igraph_biguint_sub referenced in function main [tests\test_biguint.vcxproj]
biguint.obj igraph_biguint_mul referenced in function main [tests\test_biguint.vcxproj]
biguint.obj igraph_biguint_div referenced in function main [tests\test_biguint.vcxproj]
biguint.obj igraph_biguint_print referenced in function main [tests\test_biguint.vcxproj]
```

**Comment by szhorvat** [comment]
Maybe we should also update the AppVeyor tests to test those situations which we have recently fixed, especially on Windows, which none of us uses daily.

**Comment by vtraag** [comment]
I think it would be good if we setup the full test matrix: three OS, two build types and two library usage (vendored/external).

**Comment by ntamas** [comment]
I think that the total build time for the full test matrix would probably exceed the limits of the free tiers of Appveyor and Travis; we might want to look into getting a paid account.

**Comment by szhorvat** [comment]
There are also other relevant variables to consider: currently we have a build which uses x87 instructions instead of SSE (it caught failures in the past) and a build which uses musl instead of glibc (igraph didn't work with musl in the past). In my personal fork I ran the tests on different CPU architectures, and this again has (surprisingly) caught issues in the past. Then there's 32-bit, which we don't (can't?) test at the moment. LTO is yet another thing which can potentially catch failures: I have never been able to make it work with MinGW (wonder if it works now after all the fixes), and recently I found multiple issues (fixed now) by enabling LTO on Linux.

My point is that if we take all variables into account, there is no way we can afford to cover the entire matrix (all combinations). Therefore, it's probably a better approach to try to find a balanced mix of situations to test.

That said, I am completely in favour of expanding testing!

**Comment by vtraag** [comment]
Azure Pipelines offers 10 parallel builds with unlimited minutes for open source projects, see https://azure.microsoft.com/en-us/services/devops/pipelines/?nav=min. If we do three different OS and two different build types in parallel, then we can still do some additional variations within each build. I believe Azure Pipelines is also quite fast.

**Comment by szhorvat** [comment]
I think this is done, closing.
